# **NYC Property Sales Project**
María Victoria Cabrera

[Used Dataset](https://www.kaggle.com/datasets/new-york-city/nyc-property-sales/data)

**Context**: This dataset is a record of every building or building unit (apartment, etc.) sold in the New York City property market over a 12-month period.

**Objective**: The goal is to predict the selling price of future properties in NYC.

**Notes and Observations**:
* Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement.
* This dataset uses the financial definition of a building/building unit, for tax purposes. In case a single entity owns the building in question, a sale covers the value of the entire building. In case a building is owned piecemeal by its residents (a condominium), a sale refers to a single apartment (or group of apartments) owned by some individual.

**Dictionary**:
* BOROUGH: Borough code (1 through 5).
* NEIGHBORHOOD: Neighborhood where the property is located.
* BUILDING CLASS CATEGORY: Category of the building based on its use.
* TAX CLASS AT PRESENT: Tax class as per the current assessment.
> * Class 1: Most residential property of up to three units (family homes and small stores or offices with one or two apartments attached), and most condominiums that are not more than three stories.
> * Class 2: All other property that is not in Class 1 and is primarily residential (rentals, cooperatives and condominiums). Class 2 includes:
    * Sub-Class 2a  (4 -  6 unit rental building);
    * Sub-Class 2b  (7 - 10 unit rental building);
    * Sub-Class 2c  (2 - 10 unit cooperative or condominium); and
    * Class 2  (11 units or more).
> * Class 3: Most utility property.
> * Class 4: All commercial and industrial properties, such as office, retail, factory buildings and all other properties not included in tax classes 1, 2 or 3.
* BLOCK: Tax block that the property is located on.
* LOT: Tax lot that the property is located on.
* EASE-MENT: Easement information.
* BUILDING CLASS AT PRESENT: Building class as per the current assessment.
* ADDRESS: Street address of the property.
* APARTMENT NUMBER: Apartment number (if applicable).
* ZIP CODE: Zip code of the property.
* RESIDENTIAL UNITS: Number of residential units in the building.
* COMMERCIAL UNITS: Number of commercial units in the building.
* TOTAL UNITS: Total number of units in the building.
* LAND SQUARE FEET: Land area of the property in square feet.
* GROSS SQUARE FEET: Total area of the property including both indoor and outdoor space.
* YEAR BUILT: Year the building was constructed.
* TAX CLASS AT TIME OF SALE: Tax class at the time of sale.
* BUILDING CLASS AT TIME OF SALE: Building class at the time of sale.
* SALE PRICE: Selling price of the property.
* SALE DATE: Date when the property was sold.


#Exploration of the Dataset

## Importing Libraries

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import numpy as np

## Data Loading

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Curso Data Science/Potenciales data sets TP final/Marketing y Ventas/Proyecto NYC Prices/nyc-rolling-sales.csv')

## Exploratory Analysis

In [ ]:
df.shape

(84548, 22)

In [ ]:
df.columns

Index(['Unnamed: 0', 'BOROUGH', 'NEIGHBORHOOD', 'BUILDING CLASS CATEGORY',
       'TAX CLASS AT PRESENT', 'BLOCK', 'LOT', 'EASE-MENT',
       'BUILDING CLASS AT PRESENT', 'ADDRESS', 'APARTMENT NUMBER', 'ZIP CODE',
       'RESIDENTIAL UNITS', 'COMMERCIAL UNITS', 'TOTAL UNITS',
       'LAND SQUARE FEET', 'GROSS SQUARE FEET', 'YEAR BUILT',
       'TAX CLASS AT TIME OF SALE', 'BUILDING CLASS AT TIME OF SALE',
       'SALE PRICE', 'SALE DATE'],
      dtype='object')

In [ ]:
df.head()

Unnamed: 0  BOROUGH   NEIGHBORHOOD  \
0           4        1  ALPHABET CITY   
1           5        1  ALPHABET CITY   
2           6        1  ALPHABET CITY   
3           7        1  ALPHABET CITY   
4           8        1  ALPHABET CITY   

                       BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  BLOCK  \
0  07 RENTALS - WALKUP APARTMENTS                                2A    392   
1  07 RENTALS - WALKUP APARTMENTS                                 2    399   
2  07 RENTALS - WALKUP APARTMENTS                                 2    399   
3  07 RENTALS - WALKUP APARTMENTS                                2B    402   
4  07 RENTALS - WALKUP APARTMENTS                                2A    404   

   LOT EASE-MENT BUILDING CLASS AT PRESENT                 ADDRESS  ...  \
0    6                                  C2            153 AVENUE B  ...   
1   26                                  C7   234 EAST 4TH   STREET  ...   
2   39                                  C7   197 EAST 3RD   STREET  ...   
3   21                                  C4     154 EAST 7TH STREET  ...   
4   55                                  C2  301 EAST 10TH   STREET  ...   

  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
0                 5                 0            5              1633   
1                28                 3           31              4616   
2                16                 1           17              2212   
3                10                 0           10              2272   
4                 6                 0            6              2369   

   GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
0               6440       1900                         2   
1              18690       1900                         2   
2               7803       1900                         2   
3               6794       1913                         2   
4               4615       1900                         2   

   BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
0                              C2     6625000  2017-07-19 00:00:00  
1                              C7         -    2016-12-14 00:00:00  
2                              C7         -    2016-12-09 00:00:00  
3                              C4     3936272  2016-09-23 00:00:00  
4                              C2     8000000  2016-11-17 00:00:00  

[5 rows x 22 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 22 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   Unnamed: 0                      84548 non-null  int64 
 1   BOROUGH                         84548 non-null  int64 
 2   NEIGHBORHOOD                    84548 non-null  object
 3   BUILDING CLASS CATEGORY         84548 non-null  object
 4   TAX CLASS AT PRESENT            84548 non-null  object
 5   BLOCK                           84548 non-null  int64 
 6   LOT                             84548 non-null  int64 
 7   EASE-MENT                       84548 non-null  object
 8   BUILDING CLASS AT PRESENT       84548 non-null  object
 9   ADDRESS                         84548 non-null  object
 10  APARTMENT NUMBER                84548 non-null  object
 11  ZIP CODE                        84548 non-null  int64 
 12  RESIDENTIAL UNITS               84548 non-null

In [ ]:
df.describe()

Unnamed: 0       BOROUGH         BLOCK           LOT      ZIP CODE  \
count  84548.000000  84548.000000  84548.000000  84548.000000  84548.000000   
mean   10344.359878      2.998758   4237.218976    376.224015  10731.991614   
std     7151.779436      1.289790   3568.263407    658.136814   1290.879147   
min        4.000000      1.000000      1.000000      1.000000      0.000000   
25%     4231.000000      2.000000   1322.750000     22.000000  10305.000000   
50%     8942.000000      3.000000   3311.000000     50.000000  11209.000000   
75%    15987.250000      4.000000   6281.000000   1001.000000  11357.000000   
max    26739.000000      5.000000  16322.000000   9106.000000  11694.000000   

       RESIDENTIAL UNITS  COMMERCIAL UNITS   TOTAL UNITS    YEAR BUILT  \
count       84548.000000      84548.000000  84548.000000  84548.000000   
mean            2.025264          0.193559      2.249184   1789.322976   
std            16.721037          8.713183     18.972584    537.344993   
min             0.000000          0.000000      0.000000      0.000000   
25%             0.000000          0.000000      1.000000   1920.000000   
50%             1.000000          0.000000      1.000000   1940.000000   
75%             2.000000          0.000000      2.000000   1965.000000   
max          1844.000000       2261.000000   2261.000000   2017.000000   

       TAX CLASS AT TIME OF SALE  
count               84548.000000  
mean                    1.657485  
std                     0.819341  
min                     1.000000  
25%                     1.000000  
50%                     2.000000  
75%                     2.000000  
max                     4.000000

# Preparation and Transformation

## Numeric Variables

Looking at the dataset at first glance and enquiring about the datatypes we have here, looks like we need to convert few variables into numeric.

For example sale price, land square feet, gross square feet should be numeric type.

In [ ]:
numeric_variables = ['SALE PRICE','LAND SQUARE FEET','GROSS SQUARE FEET','YEAR BUILT']

for i in numeric_variables:
    df[i] = pd.to_numeric(df[i], errors='coerce')

In [ ]:
df.dtypes

Unnamed: 0                          int64
BOROUGH                             int64
NEIGHBORHOOD                       object
BUILDING CLASS CATEGORY            object
TAX CLASS AT PRESENT               object
BLOCK                               int64
LOT                                 int64
EASE-MENT                          object
BUILDING CLASS AT PRESENT          object
ADDRESS                            object
APARTMENT NUMBER                   object
ZIP CODE                            int64
RESIDENTIAL UNITS                   int64
COMMERCIAL UNITS                    int64
TOTAL UNITS                         int64
LAND SQUARE FEET                  float64
GROSS SQUARE FEET                 float64
YEAR BUILT                          int64
TAX CLASS AT TIME OF SALE           int64
BUILDING CLASS AT TIME OF SALE     object
SALE PRICE                        float64
SALE DATE                          object
dtype: object

## Detection and Handling of Unnecessary Values

### **Irrelevant Data**

The column 'Unnamed: 0' is an iterator and is of no use here.

In [ ]:
df=df.drop(columns=['Unnamed: 0'])

In [ ]:
df=df.drop(columns=['ADDRESS'])
df=df.drop(columns=['APARTMENT NUMBER'])
df=df.drop(columns=['LOT'])
df=df.drop(columns=['BLOCK'])

While these columns provide detailed location information, they might not be directly relevant to predicting the selling price. We already have the BOROUGH and NEIGHBORHOOD columns, which should capture the location aspect.

We are not interested in overall trends or temporal patterns so we drop the sale date.

In [ ]:
df=df.drop(columns=['SALE DATE'])

 We remove these columns as they are highly correlated with TAX CLASS AT TIME OF SALE and BUILDING CLASS AT TIME OF SALE.

In [ ]:
df=df.drop(columns=['BUILDING CLASS AT PRESENT'])
df=df.drop(columns=['TAX CLASS AT PRESENT'])

### **Duplicates**

In [ ]:
df[df.duplicated()]

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
33           1  ALPHABET CITY  09 COOPS - WALKUP APARTMENTS                  
34           1  ALPHABET CITY  09 COOPS - WALKUP APARTMENTS                  
38           1  ALPHABET CITY  09 COOPS - WALKUP APARTMENTS                  
49           1  ALPHABET CITY  09 COOPS - WALKUP APARTMENTS                  
64           1  ALPHABET CITY  09 COOPS - WALKUP APARTMENTS                  
...        ...            ...                                          ...   
84414        5    WILLOWBROOK  44 CONDO PARKING                              
84415        5    WILLOWBROOK  44 CONDO PARKING                              
84416        5    WILLOWBROOK  44 CONDO PARKING                              
84445        5        WOODROW  01 ONE FAMILY DWELLINGS                       
84506        5        WOODROW  02 TWO FAMILY DWELLINGS                       

      EASE-MENT  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
33                  10009                  0                 0            0   
34                  10009                  0                 0            0   
38                  10009                  0                 0            0   
49                  10009                  0                 0            0   
64                  10009                  0                 0            0   
...         ...       ...                ...               ...          ...   
84414               10314                  0                 0            1   
84415               10314                  0                 0            1   
84416               10314                  0                 0            1   
84445               10309                  1                 0            1   
84506               10309                  2                 0            2   

       LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
33                  NaN                NaN        1920   
34                  NaN                NaN        1920   
38                  NaN                NaN        1910   
49                  NaN                NaN        1920   
64                  NaN                NaN        1920   
...                 ...                ...         ...   
84414               NaN                NaN        2015   
84415               NaN                NaN        2015   
84416               NaN                NaN        2015   
84445            2773.0             1650.0        1999   
84506            6000.0             2730.0        1976   

       TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  SALE PRICE  
33                             2                             C6    450000.0  
34                             2                             C6    450000.0  
38                             2                             C6    300000.0  
49                             2                             C6         NaN  
64                             2                             C6    490000.0  
...                          ...                            ...         ...  
84414                          4                             RG         NaN  
84415                          4                             RG         NaN  
84416                          4                             RG         NaN  
84445                          1                             A5    560000.0  
84506                          1                             B2         NaN  

[8619 rows x 14 columns]

In [ ]:
df.drop_duplicates()

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
1            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
2            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
3            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
4            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
...        ...            ...                                          ...   
84543        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84544        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84545        5        WOODROW  02 TWO FAMILY DWELLINGS                       
84546        5        WOODROW  22 STORE BUILDINGS                            
84547        5        WOODROW  35 INDOOR PUBLIC AND CULTURAL FACILITIES      

      EASE-MENT  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
0                   10009                  5                 0            5   
1                   10009                 28                 3           31   
2                   10009                 16                 1           17   
3                   10009                 10                 0           10   
4                   10009                  6                 0            6   
...         ...       ...                ...               ...          ...   
84543               10309                  2                 0            2   
84544               10309                  2                 0            2   
84545               10309                  2                 0            2   
84546               10309                  0                 7            7   
84547               10309                  0                 1            1   

       LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
0                1633.0             6440.0        1900   
1                4616.0            18690.0        1900   
2                2212.0             7803.0        1900   
3                2272.0             6794.0        1913   
4                2369.0             4615.0        1900   
...                 ...                ...         ...   
84543            2400.0             2575.0        1998   
84544            2498.0             2377.0        1998   
84545            4000.0             1496.0        1925   
84546          208033.0            64117.0        2001   
84547           10796.0             2400.0        2006   

       TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  SALE PRICE  
0                              2                             C2   6625000.0  
1                              2                             C7         NaN  
2                              2                             C7         NaN  
3                              2                             C4   3936272.0  
4                              2                             C2   8000000.0  
...                          ...                            ...         ...  
84543                          1                             B9    450000.0  
84544                          1                             B9    550000.0  
84545                          1                             B2    460000.0  
84546                          4                             K6  11693337.0  
84547                          4                             P9     69300.0  

[75929 rows x 14 columns]

### **Repetitive and uninformative columns.**

In [ ]:
num_rows = len(df)

for col in df.columns:
    cnts = df[col].value_counts(dropna=False)
    top_pct = (cnts/num_rows).iloc[0]

    if top_pct > 0.999:
        print('{0}: {1:.2f}%'.format(col, top_pct*100))
        print(cnts)
        print()

EASE-MENT: 100.00%
EASE-MENT
     84548
Name: count, dtype: int64



The column 'EASE-MENT' is blank and can be dropped.

In [ ]:
df.drop('EASE-MENT', axis=1, inplace=True)

## Detection and Handling of Inconsistent values

In [ ]:
df.describe().T

count          mean           std  min  \
BOROUGH                    84548.0  2.998758e+00  1.289790e+00  1.0   
ZIP CODE                   84548.0  1.073199e+04  1.290879e+03  0.0   
RESIDENTIAL UNITS          84548.0  2.025264e+00  1.672104e+01  0.0   
COMMERCIAL UNITS           84548.0  1.935587e-01  8.713183e+00  0.0   
TOTAL UNITS                84548.0  2.249184e+00  1.897258e+01  0.0   
LAND SQUARE FEET           58296.0  3.941676e+03  4.198397e+04  0.0   
GROSS SQUARE FEET          56936.0  4.045707e+03  3.503249e+04  0.0   
YEAR BUILT                 84548.0  1.789323e+03  5.373450e+02  0.0   
TAX CLASS AT TIME OF SALE  84548.0  1.657485e+00  8.193412e-01  1.0   
SALE PRICE                 69987.0  1.276456e+06  1.140526e+07  0.0   

                                 25%       50%       75%           max  
BOROUGH                         2.00       3.0       4.0  5.000000e+00  
ZIP CODE                    10305.00   11209.0   11357.0  1.169400e+04  
RESIDENTIAL UNITS               0.00       1.0       2.0  1.844000e+03  
COMMERCIAL UNITS                0.00       0.0       0.0  2.261000e+03  
TOTAL UNITS                     1.00       1.0       2.0  2.261000e+03  
LAND SQUARE FEET             1650.00    2325.0    3500.0  4.252327e+06  
GROSS SQUARE FEET            1046.75    1680.0    2560.0  3.750565e+06  
YEAR BUILT                   1920.00    1940.0    1965.0  2.017000e+03  
TAX CLASS AT TIME OF SALE       1.00       2.0       2.0  4.000000e+00  
SALE PRICE                 225000.00  530000.0  950000.0  2.210000e+09

### **Dealing with the '0' values in columns**

*Sale Price:*

In [ ]:
rows_with_zero_sale_price = df[df['SALE PRICE'] == 0]
rows_with_zero_sale_price

BOROUGH     NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
18311        2         BATHGATE  01 ONE FAMILY DWELLINGS                       
18313        2         BATHGATE  01 ONE FAMILY DWELLINGS                       
18314        2         BATHGATE  02 TWO FAMILY DWELLINGS                       
18315        2         BATHGATE  02 TWO FAMILY DWELLINGS                       
18320        2         BATHGATE  02 TWO FAMILY DWELLINGS                       
...        ...              ...                                          ...   
49391        3  WYCKOFF HEIGHTS  27 FACTORIES                                  
49396        3  WYCKOFF HEIGHTS  31 COMMERCIAL VACANT LAND                     
49398        3  WYCKOFF HEIGHTS  41 TAX CLASS 4 - OTHER                        
49399        3  WYCKOFF HEIGHTS  41 TAX CLASS 4 - OTHER                        
49401        3  WYCKOFF HEIGHTS  44 CONDO PARKING                              

       ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
18311     10457                  1                 0            1   
18313     10458                  1                 0            1   
18314     10457                  2                 0            2   
18315     10457                  2                 0            2   
18320     10457                  2                 0            2   
...         ...                ...               ...          ...   
49391     11237                  0                 1            1   
49396     11237                  0                 0            0   
49398     11237                  0                 0            0   
49399     11237                  0                 0            0   
49401     11237                  0                 0            1   

       LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
18311            1209.0             1048.0        1901   
18313             922.0             1248.0        1901   
18314            2000.0             2400.0        1993   
18315            2361.0             2394.0        1995   
18320            1790.0             2340.0        1998   
...                 ...                ...         ...   
49391           10000.0            10000.0        1931   
49396            2500.0                0.0           0   
49398            2415.0                0.0           0   
49399            2390.0                0.0           0   
49401               0.0                0.0           0   

       TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  SALE PRICE  
18311                          1                             A1         0.0  
18313                          1                             A1         0.0  
18314                          1                             B1         0.0  
18315                          1                             B1         0.0  
18320                          1                             B1         0.0  
...                          ...                            ...         ...  
49391                          4                             F4         0.0  
49396                          4                             V1         0.0  
49398                          4                             Z9         0.0  
49399                          4                             Z9         0.0  
49401                          4                             RP         0.0  

[10228 rows x 13 columns]

In [ ]:
df_nonzero_sale_price = df[df['SALE PRICE'] != 0]
df = df_nonzero_sale_price.dropna()

We dropped 10228 rows with SALE PRICE = 0

*Total Units:*

In [ ]:
rows_with_zero_total_units = df[df['TOTAL UNITS'] == 0]
rows_with_zero_total_units

BOROUGH             NEIGHBORHOOD  \
5603         1           HARLEM-CENTRAL   
17503        1  UPPER WEST SIDE (79-96)   
18354        2                 BATHGATE   
18355        2                 BATHGATE   
18356        2                 BATHGATE   
...        ...                      ...   
71103        4                RIDGEWOOD   
71104        4                RIDGEWOOD   
81560        5               PORT IVORY   
82013        5              PRINCES BAY   
82569        5     ROSSVILLE-PORT MOBIL   

                           BUILDING CLASS CATEGORY  ZIP CODE  \
5603   37 RELIGIOUS FACILITIES                         10035   
17503  33 EDUCATIONAL FACILITIES                       10024   
18354  10 COOPS - ELEVATOR APARTMENTS                  10457   
18355  10 COOPS - ELEVATOR APARTMENTS                  10457   
18356  10 COOPS - ELEVATOR APARTMENTS                  10457   
...                                            ...       ...   
71103  38 ASYLUMS AND HOMES                            11385   
71104  38 ASYLUMS AND HOMES                            11385   
81560  06 TAX CLASS 1 - OTHER                          10303   
82013  21 OFFICE BUILDINGS                             10309   
82569  34 THEATRES                                     10309   

       RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
5603                   0                 0            0            5973.0   
17503                  0                 0            0            2554.0   
18354                  0                 0            0               0.0   
18355                  0                 0            0               0.0   
18356                  0                 0            0               0.0   
...                  ...               ...          ...               ...   
71103                  0                 0            0            2650.0   
71104                  0                 0            0            2650.0   
81560                  0                 0            0            2625.0   
82013                  0                 0            0            7160.0   
82569                  0                 0            0         3014056.0   

       GROSS SQUARE FEET  YEAR BUILT  TAX CLASS AT TIME OF SALE  \
5603             24626.0        1900                          4   
17503            15000.0        1910                          4   
18354                0.0        1935                          2   
18355                0.0        1935                          2   
18356                0.0        1941                          2   
...                  ...         ...                        ...   
71103             5600.0        1931                          4   
71104             5600.0        1931                          4   
81560              600.0        1931                          1   
82013             1084.0        1931                          4   
82569           349503.0        1971                          4   

      BUILDING CLASS AT TIME OF SALE  SALE PRICE  
5603                              M9    330000.0  
17503                             W8        10.0  
18354                             D4     40000.0  
18355                             D4     40000.0  
18356                             D4     60000.0  
...                              ...         ...  
71103                             N9   2550000.0  
71104                             N9   1450000.0  
81560                             G0    140000.0  
82013                             O1    420000.0  
82569                             J9   7211750.0  

[4156 rows x 13 columns]

In [ ]:
df_nonzero_total_units = df[df['TOTAL UNITS'] != 0]
df = df_nonzero_total_units.dropna()

We dropped 4156 rows with 0 Total Units.

*Land square feet:*

In [ ]:
df = df[df['LAND SQUARE FEET']!=0].copy()

Keeping only the rows which have land square feet != 0.

*Gross square feet:*

In [ ]:
df[df['GROSS SQUARE FEET']==0]

BOROUGH            NEIGHBORHOOD  \
20039        2            EAST TREMONT   
20188        2                 FORDHAM   
21494        2     MORRISANIA/LONGWOOD   
21765        2  MOTT HAVEN/PORT MORRIS   
21837        2   MOUNT HOPE/MOUNT EDEN   
21839        2   MOUNT HOPE/MOUNT EDEN   
24357        2             THROGS NECK   
24358        2             THROGS NECK   
24359        2             THROGS NECK   
24448        2             THROGS NECK   
25089        2          WILLIAMSBRIDGE   
25504        3              BATH BEACH   
25521        3              BATH BEACH   
26573        3      BEDFORD STUYVESANT   
27274        3      BEDFORD STUYVESANT   
27376        3      BEDFORD STUYVESANT   
28906        3             BOERUM HILL   
29510        3            BOROUGH PARK   
29930        3            BOROUGH PARK   
30165        3          BRIGHTON BEACH   
30227        3          BRIGHTON BEACH   
31053        3        BROOKLYN HEIGHTS   
31054        3        BROOKLYN HEIGHTS   
31055        3        BROOKLYN HEIGHTS   
32202        3                CANARSIE   
32203        3                CANARSIE   
32485        3                CANARSIE   
33943        3            CONEY ISLAND   
33946        3            CONEY ISLAND   
34386        3           CROWN HEIGHTS   
34744        3           CROWN HEIGHTS   
34745        3           CROWN HEIGHTS   
34874        3           CYPRESS HILLS   
35069        3           CYPRESS HILLS   
35070        3           CYPRESS HILLS   
36954        3           EAST NEW YORK   
36959        3           EAST NEW YORK   
37448        3        FLATBUSH-CENTRAL   
38371        3           FLATBUSH-EAST   
40306        3               GRAVESEND   
40309        3               GRAVESEND   
40356        3               GRAVESEND   
40740        3              GREENPOINT   
41019        3              GREENPOINT   
41096        3             JAMAICA BAY   
43506        3     OCEAN PARKWAY-NORTH   
43994        3     OCEAN PARKWAY-SOUTH   
44110        3     OCEAN PARKWAY-SOUTH   
45672        3                RED HOOK   
45676        3                RED HOOK   
45677        3                RED HOOK   
45678        3                RED HOOK   
48204        3       WILLIAMSBURG-EAST   
48693        3      WILLIAMSBURG-SOUTH   
48694        3      WILLIAMSBURG-SOUTH   

                           BUILDING CLASS CATEGORY  ZIP CODE  \
20039  05 TAX CLASS 1 VACANT LAND                      10457   
20188  05 TAX CLASS 1 VACANT LAND                      10458   
21494  05 TAX CLASS 1 VACANT LAND                      10455   
21765  29 COMMERCIAL GARAGES                           10454   
21837  03 THREE FAMILY DWELLINGS                       10457   
21839  05 TAX CLASS 1 VACANT LAND                      10457   
24357  02 TWO FAMILY DWELLINGS                         10465   
24358  02 TWO FAMILY DWELLINGS                         10465   
24359  02 TWO FAMILY DWELLINGS                         10465   
24448  29 COMMERCIAL GARAGES                           10465   
25089  02 TWO FAMILY DWELLINGS                         10466   
25504  03 THREE FAMILY DWELLINGS                       11214   
25521  03 THREE FAMILY DWELLINGS                       11214   
26573  02 TWO FAMILY DWELLINGS                         11221   
27274  05 TAX CLASS 1 VACANT LAND                      11221   
27376  07 RENTALS - WALKUP APARTMENTS                  11233   
28906  31 COMMERCIAL VACANT LAND                       11217   
29510  03 THREE FAMILY DWELLINGS                       11219   
29930  14 RENTALS - 4-10 UNIT                          11219   
30165  01 ONE FAMILY DWELLINGS                         11235   
30227  05 TAX CLASS 1 VACANT LAND                      11235   
31053  31 COMMERCIAL VACANT LAND                       11201   
31054  31 COMMERCIAL VACANT LAND                       11201   
31055  31 COMMERCIAL VACANT LAND                       11201   
32202  01 ONE FAMILY DWELLINGS                        

In [ ]:
df.loc[df['GROSS SQUARE FEET'] == 0, 'GROSS SQUARE FEET'] = df['LAND SQUARE FEET']

 We replaced the rows which have 0 value in gross square feet with corresponding land square feet, considering they have no additional structures on the land and exterior surfaces.

*Total units:*

In [ ]:
df = df[df['TOTAL UNITS']>0]

Keeping only those rows which have residential units + commercial units equal to total units.

In [ ]:
df = df[df['RESIDENTIAL UNITS'] + df['COMMERCIAL UNITS'] == df['TOTAL UNITS']].copy()

*Year built:*

The column YEAR BUILT also has 0 values.
After some research, we discovered the oldest property in nyc was built in 1652 (https://secretnyc.co/the-wyckoff-house-oldest-nyc-building/#:~:text=Standing%20strong%20for%20over%20three,oldest%20surviving%20building%20in%20NYC.&text=Built%20in%201652%2C%20The%20Wyckoff,Claesen%20and%20his%20wife%20Grietje.), so we are keeping only those properties that were built after 1650.

In [ ]:
df = df[df['YEAR BUILT']>1650]

*Zip code:*

In [ ]:
df = df[df['ZIP CODE']!=0].copy()

Keeping only the rows with ZIP CODE != 0

### **Inconsistent capitalization**

In [ ]:
object_columns = df.select_dtypes(include=['object']).columns

for columns in object_columns:
    print(f"\nUnique values in '{columns}':")
    print(df[columns].value_counts(dropna=False))


Unique values in 'NEIGHBORHOOD':
FLUSHING-NORTH              781
BEDFORD STUYVESANT          585
EAST NEW YORK               574
GREAT KILLS                 468
FLATBUSH-EAST               454
                           ... 
BRONX PARK                    1
CITY ISLAND-PELHAM STRIP      1
EAST RIVER                    1
SOUTHBRIDGE                   1
ROOSEVELT ISLAND              1
Name: NEIGHBORHOOD, Length: 249, dtype: int64

Unique values in 'BUILDING CLASS CATEGORY':
01 ONE FAMILY DWELLINGS                         12524
02 TWO FAMILY DWELLINGS                          9648
03 THREE FAMILY DWELLINGS                        2281
07 RENTALS - WALKUP APARTMENTS                   1721
22 STORE BUILDINGS                                469
14 RENTALS - 4-10 UNIT                            316
21 OFFICE BUILDINGS                               210
08 RENTALS - ELEVATOR APARTMENTS                  196
29 COMMERCIAL GARAGES                             166
30 WAREHOUSES                        

## Detection and handling of outliers.

In [ ]:
df.kurt(numeric_only=True)[:10]

BOROUGH                         -0.302024
ZIP CODE                        -1.187292
RESIDENTIAL UNITS             3112.540865
COMMERCIAL UNITS             23126.144497
TOTAL UNITS                   3866.662264
LAND SQUARE FEET              7915.450022
GROSS SQUARE FEET             5415.120078
YEAR BUILT                      -0.059279
TAX CLASS AT TIME OF SALE       10.218741
SALE PRICE                   10380.103783
dtype: float64

The column COMMERCIAL UNITS is the one with the highest kurtosis.

In [ ]:
df['COMMERCIAL UNITS'].describe()

count    29299.00000
mean         0.32868
std         14.02409
min          0.00000
25%          0.00000
50%          0.00000
75%          0.00000
max       2261.00000
Name: COMMERCIAL UNITS, dtype: float64

In [ ]:
fig = px.box(df, x='COMMERCIAL UNITS')
fig.show()

In [ ]:
df[df['COMMERCIAL UNITS'] == 2261]

BOROUGH NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
64831        4      JAMAICA  33 EDUCATIONAL FACILITIES                     

       ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
64831     11432                  0              2261         2261   

       LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
64831            2940.0             4522.0        1920   

       TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  SALE PRICE  
64831                          4                             W9    315000.0

We found that the building with 2261 commercial units was an educational facility so we leave it as it is.

SALE PRICE was the column with the second highest kurtosis

In [ ]:
df['SALE PRICE'].describe()

count    2.929900e+04
mean     1.636901e+06
std      1.700185e+07
min      1.000000e+00
25%      4.200000e+05
50%      6.200000e+05
75%      9.500000e+05
max      2.210000e+09
Name: SALE PRICE, dtype: float64

In [ ]:
fig = px.box(df, x='SALE PRICE')
fig.show()

Reviewing the plot, it initially appears that properties with a sale price of 2 billion are outliers. However, upon closer examination, it was found that this specific property is valid, as detailed on its Wikipedia page (https://en.wikipedia.org/wiki/245_Park_Avenue).


As a decision, I have chosen to retain rows with sale prices greater than USD 50,000 and less than USD 500,000,000. This selection criterion is based on the assumption that sale prices below USD 50,000 might need replacement or can be dropped. Conversely, there are only a few instances where the sale price exceeds 500,000,000 USD (e.g., the most expensive property in the original dataset is around 2 billion USD). These high values can be considered outliers and have the potential to impact models applied to the data, potentially altering predictions made on the target variable.

In [ ]:
df= df[(df['SALE PRICE']>50000) & (df['SALE PRICE']<500000000)].copy()

## Detection of Missing Data

In [ ]:
df.isna().sum()

BOROUGH                           0
NEIGHBORHOOD                      0
BUILDING CLASS CATEGORY           0
ZIP CODE                          0
RESIDENTIAL UNITS                 0
COMMERCIAL UNITS                  0
TOTAL UNITS                       0
LAND SQUARE FEET                  0
GROSS SQUARE FEET                 0
YEAR BUILT                        0
TAX CLASS AT TIME OF SALE         0
BUILDING CLASS AT TIME OF SALE    0
SALE PRICE                        0
dtype: int64

We have not found any null values.

**Resetting the index**

In [ ]:
df.reset_index(drop = True)

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
1            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
2            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
3            1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
4            1  ALPHABET CITY  08 RENTALS - ELEVATOR APARTMENTS              
...        ...            ...                                          ...   
28096        5        WOODROW  02 TWO FAMILY DWELLINGS                       
28097        5        WOODROW  02 TWO FAMILY DWELLINGS                       
28098        5        WOODROW  02 TWO FAMILY DWELLINGS                       
28099        5        WOODROW  22 STORE BUILDINGS                            
28100        5        WOODROW  35 INDOOR PUBLIC AND CULTURAL FACILITIES      

       ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
0         10009                  5                 0            5   
1         10009                 10                 0           10   
2         10009                  6                 0            6   
3         10009                  8                 0            8   
4         10009                 24                 0           24   
...         ...                ...               ...          ...   
28096     10309                  2                 0            2   
28097     10309                  2                 0            2   
28098     10309                  2                 0            2   
28099     10309                  0                 7            7   
28100     10309                  0                 1            1   

       LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
0                1633.0             6440.0        1900   
1                2272.0             6794.0        1913   
2                2369.0             4615.0        1900   
3                1750.0             4226.0        1920   
4                4489.0            18523.0        1920   
...                 ...                ...         ...   
28096            2400.0             2575.0        1998   
28097            2498.0             2377.0        1998   
28098            4000.0             1496.0        1925   
28099          208033.0            64117.0        2001   
28100           10796.0             2400.0        2006   

       TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  SALE PRICE  
0                              2                             C2   6625000.0  
1                              2                             C4   3936272.0  
2                              2                             C2   8000000.0  
3                              2                             C4   3192840.0  
4                              2                             D9  16232000.0  
...                          ...                            ...         ...  
28096                          1                             B9    450000.0  
28097                          1                             B9    550000.0  
28098                          1                             B2    460000.0  
28099                          4                             K6  11693337.0  
28100                          4                             P9     69300.0  

[28101 rows x 13 columns]

In [ ]:
df.to_csv(r'/content/drive/MyDrive/Curso Data Science/Potenciales data sets TP final/Marketing y Ventas/Proyecto NYC Prices/nyc_clean_dataset1.csv', index = False)

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Curso Data Science/Potenciales data sets TP final/Marketing y Ventas/Proyecto NYC Prices/nyc_clean_dataset1.csv')

# Data Visualization Phase

In [ ]:
avg_sale_prices = df.groupby('NEIGHBORHOOD')['SALE PRICE'].mean()

fig = px.bar(avg_sale_prices, x=avg_sale_prices.index, y='SALE PRICE', title='Average Sale Prices by Neighborhood', text_auto=True)

fig.update_layout(xaxis_tickangle=-45)

fig.show()

In [ ]:
avg_sale_prices = df.groupby('BOROUGH')['SALE PRICE'].mean()

fig = px.bar(avg_sale_prices, x=avg_sale_prices.index, y='SALE PRICE', title='Average Sale Prices by Borough', text_auto=True)


fig.show()

In [ ]:
avg_sale_price_by_units = df.groupby('TOTAL UNITS')['SALE PRICE'].mean().reset_index()

fig = px.scatter(avg_sale_price_by_units, x='TOTAL UNITS', y='SALE PRICE',
                 labels={'TOTAL UNITS': 'Total Units', 'SALE PRICE': 'Average Sale Price'},
                 title='Average Sale Price by Total Units',
                 width=800, height=600,
                 color_discrete_sequence=['#171bb0'])

fig.update_layout(xaxis_title='Total Units',
                  yaxis_title='Average Sale Price',
                  font=dict(size=12))

fig.show()

In [ ]:
avg_sale_price_by_year = df.groupby('YEAR BUILT')['SALE PRICE'].mean().reset_index()

fig = px.line(avg_sale_price_by_year, x='YEAR BUILT', y='SALE PRICE',
              labels={'YEAR BUILT': 'Year Built', 'SALE PRICE': 'Average Sale Price'},
              title='Average Sale Price by Year Built',
              width=800, height=600,
              line_shape="spline",
              color_discrete_sequence=['#171bb0'])

fig.update_layout(xaxis_title='Year Built',
                  yaxis_title='Average Sale Price',
                  font=dict(size=12))

fig.show()

In [ ]:
fig = px.scatter(df, x='GROSS SQUARE FEET', y='SALE PRICE',
                 labels={'GROSS SQUARE FEET': 'Gross Square Feet', 'SALE PRICE': 'Sale Price'},
                 title='Effect of Gross Square Feet on Sale Price',
                 width=800, height=600,
                 color_discrete_sequence=['#171bb0'])

fig.update_layout(xaxis_title='Gross Square Feet',
                  yaxis_title='Sale Price',
                  font=dict(size=12))

fig.show()

# Data Modeling phase

## Data preparation

### Encoding Categorical Variables

In [ ]:
pip install category-encoders

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 1.3 MB/s eta 0:00:00


In [ ]:
object_columns = df.select_dtypes(include=['object'])
print('Variables to Encode:')
print(object_columns.dtypes)

Variables to Encode:
NEIGHBORHOOD                      object
BUILDING CLASS CATEGORY           object
BUILDING CLASS AT TIME OF SALE    object
dtype: object


In [ ]:
df['NEIGHBORHOOD'].value_counts(dropna=False)

NEIGHBORHOOD
FLUSHING-NORTH              781
BEDFORD STUYVESANT          585
EAST NEW YORK               574
GREAT KILLS                 468
FLATBUSH-EAST               454
                           ... 
BRONX PARK                    1
CITY ISLAND-PELHAM STRIP      1
EAST RIVER                    1
SOUTHBRIDGE                   1
ROOSEVELT ISLAND              1
Name: count, Length: 249, dtype: int64

In [ ]:
df['BUILDING CLASS CATEGORY'].value_counts(dropna=False)

BUILDING CLASS CATEGORY
01 ONE FAMILY DWELLINGS                         12524
02 TWO FAMILY DWELLINGS                          9648
03 THREE FAMILY DWELLINGS                        2281
07 RENTALS - WALKUP APARTMENTS                   1721
22 STORE BUILDINGS                                469
14 RENTALS - 4-10 UNIT                            316
21 OFFICE BUILDINGS                               210
08 RENTALS - ELEVATOR APARTMENTS                  196
29 COMMERCIAL GARAGES                             166
30 WAREHOUSES                                     162
27 FACTORIES                                      102
37 RELIGIOUS FACILITIES                            51
26 OTHER HOTELS                                    50
10 COOPS - ELEVATOR APARTMENTS                     29
32 HOSPITAL AND HEALTH FACILITIES                  27
05 TAX CLASS 1 VACANT LAND                         24
33 EDUCATIONAL FACILITIES                          24
23 LOFT BUILDINGS                                  21
38 A

In [ ]:
df['BUILDING CLASS AT TIME OF SALE'].value_counts(dropna=False)

BUILDING CLASS AT TIME OF SALE
A1    4670
A5    3986
B2    3104
B1    2728
B3    2466
      ... 
H9       1
I3       1
J4       1
HS       1
CM       1
Name: count, Length: 126, dtype: int64

In [ ]:
# Initialize empty dictionaries to store hashed values
hashed_building_class_at_time_of_sale = {}
hashed_building_class_category = {}
hashed_neighborhood = {}

# Encode categories using hashing and store hashed values in dictionaries
for category in df['BUILDING CLASS AT TIME OF SALE'].unique():
    hashed_value = hash(category)
    hashed_building_class_at_time_of_sale[category] = hashed_value

for category in df['BUILDING CLASS CATEGORY'].unique():
    hashed_value = hash(category)
    hashed_building_class_category[category] = hashed_value

for category in df['NEIGHBORHOOD'].unique():
    hashed_value = hash(category)
    hashed_neighborhood[category] = hashed_value

# Add encoded variables back to the DataFrame
df['ENCODED_BUILDING_CLASS_AT_TIME_OF_SALE'] = df['BUILDING CLASS AT TIME OF SALE'].map(hashed_building_class_at_time_of_sale)
df['ENCODED_BUILDING_CLASS_CATEGORY'] = df['BUILDING CLASS CATEGORY'].map(hashed_building_class_category)
df['ENCODED_NEIGHBORHOOD'] = df['NEIGHBORHOOD'].map(hashed_neighborhood)

df.drop(columns=['BUILDING CLASS AT TIME OF SALE', 'BUILDING CLASS CATEGORY', 'NEIGHBORHOOD'], inplace=True)

df

BOROUGH  ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
0            1     10009                  5                 0            5   
3            1     10009                 10                 0           10   
4            1     10009                  6                 0            6   
6            1     10009                  8                 0            8   
9            1     10009                 24                 0           24   
...        ...       ...                ...               ...          ...   
84543        5     10309                  2                 0            2   
84544        5     10309                  2                 0            2   
84545        5     10309                  2                 0            2   
84546        5     10309                  0                 7            7   
84547        5     10309                  0                 1            1   

       LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  \
0                1633.0             6440.0        1900   
3                2272.0             6794.0        1913   
4                2369.0             4615.0        1900   
6                1750.0             4226.0        1920   
9                4489.0            18523.0        1920   
...                 ...                ...         ...   
84543            2400.0             2575.0        1998   
84544            2498.0             2377.0        1998   
84545            4000.0             1496.0        1925   
84546          208033.0            64117.0        2001   
84547           10796.0             2400.0        2006   

       TAX CLASS AT TIME OF SALE  SALE PRICE  \
0                              2   6625000.0   
3                              2   3936272.0   
4                              2   8000000.0   
6                              2   3192840.0   
9                              2  16232000.0   
...                          ...         ...   
84543                          1    450000.0   
84544                          1    550000.0   
84545                          1    460000.0   
84546                          4  11693337.0   
84547                          4     69300.0   

       ENCODED_BUILDING_CLASS_AT_TIME_OF_SALE  \
0                         6583763651214800611   
3                         4025873108569691218   
4                         6583763651214800611   
6                         4025873108569691218   
9                         5972083703994063748   
...                                       ...   
84543                      -60971368967281295   
84544                      -60971368967281295   
84545                    -1329648869005401101   
84546                    -6547920981096436003   
84547                     4021474049408924606   

       ENCODED_BUILDING_CLASS_CATEGORY  ENCODED_NEIGHBORHOOD  
0                 -5739492016562804284  -5332979656013783429  
3                 -5739492016562804284  -5332979656013783429  
4                 -5739492016562804284  -5332979656013783429  
6                 -5739492016562804284  -5332979656013783429  
9                 -4008231925625714032  -5332979656013783429  
...                                ...                   ...  
84543              7017841900620899357    960185785614921651  
84544              7017841900620899357    960185785614921651  
84545              7017841900620899357    960185785614921651  
84546              3558449468690594859    960185785614921651  
84547             -1761274794608789271    960185785614921651  

[28101 rows x 13 columns]

## Feature Selection

In [ ]:
y = df["SALE PRICE"]
X = df.drop(["SALE PRICE"], axis=1)

**Recursive Feature Elimination with a Random Forest Regressor**

In [ ]:
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestRegressor

rfe = RFE(RandomForestRegressor(n_estimators=100, random_state=42), n_features_to_select=8)

rfe.fit(X, y)

print("Feature ranking:", rfe.ranking_)

print("Selected features:", X.columns[rfe.support_])

Feature ranking: [5 1 3 2 1 1 1 1 4 1 1 1]
Selected features: Index(['ZIP CODE', 'TOTAL UNITS', 'LAND SQUARE FEET', 'GROSS SQUARE FEET',
       'YEAR BUILT', 'ENCODED_BUILDING_CLASS_AT_TIME_OF_SALE',
       'ENCODED_BUILDING_CLASS_CATEGORY', 'ENCODED_NEIGHBORHOOD'],
      dtype='object')


## Machine Learning Predictive Modeling

We will test three regression models' performance with the features selected above:

In [ ]:
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

from sklearn.metrics import mean_squared_error as MSE
from sklearn.metrics import r2_score

In [ ]:
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

from sklearn.metrics import mean_squared_error as MSE
from sklearn.metrics import r2_score

y = df["SALE PRICE"]
X = df[['ZIP CODE', 'TOTAL UNITS', 'LAND SQUARE FEET', 'GROSS SQUARE FEET',
       'YEAR BUILT', 'ENCODED_BUILDING_CLASS_AT_TIME_OF_SALE',
       'ENCODED_BUILDING_CLASS_CATEGORY', 'ENCODED_NEIGHBORHOOD']]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 3)

LR_model = LinearRegression()
DT_model = DecisionTreeRegressor()
RFR_model = RandomForestRegressor(n_estimators = 50)
GBR_model = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

model_list = [LR_model, DT_model, RFR_model, GBR_model]

for model in model_list:
    model.fit(X_train, y_train)
    predictions = model.predict(X_test)
    print(str(model) + " test MSE:", MSE(y_test, predictions))
    print(str(model) + " test corr:", np.corrcoef(y_test, predictions)[0,1])
    print(str(model) + " r2:", r2_score(y_test, predictions), "\n")

LinearRegression() test MSE: 69625019301741.266
LinearRegression() test corr: 0.2869830755346001
LinearRegression() r2: -0.31272489211327525 

DecisionTreeRegressor() test MSE: 38809384959000.52
DecisionTreeRegressor() test corr: 0.5763329357584975
DecisionTreeRegressor() r2: 0.26827961853057014 

RandomForestRegressor(n_estimators=50) test MSE: 25784726553515.87
RandomForestRegressor(n_estimators=50) test corr: 0.7170690793441331
RandomForestRegressor(n_estimators=50) r2: 0.5138492926451812 

GradientBoostingRegressor(random_state=42) test MSE: 25393141413194.246
GradientBoostingRegressor(random_state=42) test corr: 0.7235114012914754
GradientBoostingRegressor(random_state=42) r2: 0.5212323219963704 



Since the random forest regression model was one of the most accurate algorithms, we will try to see if the model can be improved using more trees in the forest.

In [ ]:
model = RandomForestRegressor(n_estimators = 200).fit(X_train, y_train)
predictions = model.predict(X_test)

print("MSE: ", MSE(y_test, predictions))
print("Correlation: ", np.corrcoef(y_test, predictions)[0,1])
print("R-squared: ", r2_score(y_test, predictions))

MSE:  23822677030896.605
Correlation:  0.7422555279083503
R-squared:  0.5508421907977691


Now let's compare these results with an improved version of Gradient Boosting Regression using GridSearch:

In [ ]:
from sklearn.model_selection import GridSearchCV

y = df["SALE PRICE"]
X = df[['ZIP CODE', 'TOTAL UNITS', 'LAND SQUARE FEET', 'GROSS SQUARE FEET',
       'YEAR BUILT', 'ENCODED_BUILDING_CLASS_AT_TIME_OF_SALE',
       'ENCODED_BUILDING_CLASS_CATEGORY', 'ENCODED_NEIGHBORHOOD']]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

param_grid = {'n_estimators': [50, 100, 200],
              'learning_rate': [0.01, 0.1, 1],
              'max_depth': [3, 5, 7]}
grid_search = GridSearchCV(GradientBoostingRegressor(random_state=42), param_grid, cv=5, scoring='neg_mean_squared_error')

grid_search.fit(X_train, y_train)

best_params = grid_search.best_params_
best_score = -grid_search.best_score_

print("Best Parameters: ", best_params)
print("Best Score (Negative RMSE): ", best_score)

model = GradientBoostingRegressor(**best_params, random_state=42)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = MSE(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print("Mean Squared Error: ", mse)
print("R-squared:", r2)
print("Training accuracy :", model.score(X_train, y_train))
print("Testing accuracy :", model.score(X_test, y_test))


Best Parameters:  {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 200}
Best Score (Negative RMSE):  25356464629594.605
Mean Squared Error:  17956328256301.46
R-squared: 0.6163062240211097
Training accuracy : 0.7093499892965537
Testing accuracy : 0.6163062240211097


# Conclusion

In conclusion, our data analysis revealed several key insights about the NYC property market. We identified that factors such as neighborhood, building class category, and the size of residential and commercial units have significant impacts on property sale prices. Through the use of machine learning models, including random forest regression and gradient boosting with hyperparameter optimization using Grid Search, we were able to predict property sale prices with a  moderate degree of accuracy. Our best-performing model achieved an R-squared value of 0.62, indicating 62% of the variance in the dependent variable can be explained by the independent variables included in the model.
The remaining 38% of the variance is unexplained and could be due to factors not included in the model or random variation.
Additionally, we uncovered outliers and anomalies in the data, which were carefully addressed to ensure the integrity of our analyses. Based on these findings, we recommend that stakeholders in the NYC property market consider factors such as neighborhood demographics, building classifications, and unit sizes when making pricing decisions. Further research could explore additional features or incorporate external data sources for enhanced predictive modeling. Overall, this project provides valuable insights into the dynamics of the NYC property market and lays the groundwork for informed decision-making in the real estate industry.